Hidden data streams

probakowski opened the pull request:
This change adds possibility to mark data stream as hidden. To use it you set usual `"index.hidden":true` setting in data stream template.
This change also relaxes naming convention for data streams - data stream can start with `.` but can't start with backing index prefix (`.ds-` currently) to avoid clashes.

Closes #63889

probakowski committed:
Hidden data streams

probakowski committed:
whitespace reverted

probakowski committed:
stricter ds name

probakowski committed:
Revert "stricter ds name"

This reverts commit 100dba5f3c9561a2e8bc5c52bffa5881c114b6c7.

martijnvg reviewed (commented):
The implementation looks good at first glance. Can you also add a rest test for hiding a data stream? iirc some changes are also needed in security.

probakowski committed:
String.format removed

danhermann reviewed (commented):
Does this change the behavior of the GetDataStream API? I think we'll need to add the `expand_wildcards` option to that API for consistency in the handling of hidden data streams.

probakowski committed:
fix test

probakowski committed:
fix GetDataStream action

probakowski committed:
fix test

probakowski committed:
fix test

probakowski committed:
rest test

probakowski committed:
rest test

probakowski committed:
spotless

probakowski committed:
tests

probakowski committed:
Merge remote-tracking branch 'origin/master' into hidden-data-streams

probakowski committed:
Delete a.json

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

probakowski commented:
Thanks for review, could I ask for another round?
@martijnvg do you know remember what it was about security? I've tested it with security on locally and used RBAC as well, it all worked correctly. I've added REST tests.
@danhermann for now I've updated GetDataStream API to show all data streams, the same goes for DeleteDataStream API. I think adding `expand_wildcards` param is good idea but I think it can be done in separate issue, WDYT?

probakowski commented:
@elasticmachine run elasticsearch-ci/1

probakowski commented:
@elasticmachine run elasticsearch-ci/1

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -156,9 +157,11 @@ static ClusterState createDataStream(MetadataCreateIndexService metadataCreateIn
         DataStream.TimestampField timestampField = new DataStream.TimestampField(fieldName);
-        DataStream newDataStream = new DataStream(request.name, timestampField, List.of(firstBackingIndex.getIndex()));
+        Template template = composableTemplate.template();
+        boolean hidden = template != null && template.settings() != null && IndexMetadata.INDEX_HIDDEN_SETTING.get(template.settings());
```
I feel that reusing the `index.hidden` setting is adding ambiguity to the setting, since it will then be controlling whether an index is hidden or data stream depending on the context how it is being used.

I would prefer that a data stream can be made hidden via the `data_stream` section of a composable index template. I think this way it is clearer what is being made hidden. Currently `DataStreamTemplate` is empty, but I think we will be adding more configuration options that are data stream specific later (e.g. timestamp field name). 

Note that aliases are also made hidden via a property on an alias instead of reusing the `index.hidden` setting. So making data stream hidden via a parameter on `DataStreamTemplate` is inline how aliases
can be made hidden.

Let me know what you think. I think this change would be a minor change to this PR.

probakowski replied:
That makes sense, I'll change it

probakowski replied:
It's done, now it uses `data_stream` part of a template
```json
"data_stream": {
    "hidden": true
}
```

probakowski committed:
added expand_wildcards for GetDataStream and DeleteDataStream

martijnvg reviewed (commented):
> do you remember what it was about security? 

I mixed something up. So never mind that comment.

>  I've added REST tests.

That is great. However I would prefer if also a basic yaml test is added for hidden data streams.
This way hidden data streams is tested without security in the free distribution and
with security running with a trail license.

probakowski committed:
unused imports

probakowski committed:
Merge remote-tracking branch 'origin/master' into hidden-data-streams

danhermann commented:
> @danhermann for now I've updated GetDataStream API to show all data streams, the same goes for DeleteDataStream API. I think adding `expand_wildcards` param is good idea but I think it can be done in separate issue, WDYT?

I'm fine with a separate PR for it so long as we have plans to get it in soon because I think that discrepancies in whether APIs resolve hidden data streams by default provide a poor user experience.

danhermann commented:
Also, I'm a little surprised that the spotless check doesn't catch formatting discrepancies such as the ones on the following lines which are especially noticeable since they're adjacent to several blocks of code in the same class that use our standard formatting.

https://github.com/elastic/elasticsearch/pull/63987/files#diff-94129876125bd784be1417723d5e53d9f1ccf91498e5ba7056a11201ecd686a8R83-R85

Even if spotless does not catch that, I know that you can import our [code formatting config file](https://github.com/elastic/elasticsearch/blob/master/buildSrc/formatterConfig.xml) into either IntelliJ or Eclipse and it will automatically format those lines appropriately.

probakowski committed:
add hidden setting to data stream template

probakowski committed:
fix expand_wildcards

probakowski committed:
Merge remote-tracking branch 'origin/master' into hidden-data-streams

probakowski committed:
spotless

probakowski committed:
fix compilation

probakowski committed:
unused import

probakowski committed:
yaml test

probakowski committed:
fix test

probakowski committed:
fix cleanup

probakowski commented:
Thanks for catching it @danhermann, it was oversight on my side, I have correct formatter, I just didn't used it here. Strange spotless hasn't caught it though. 
I've added `expand_wildcards` to both APIs.
@martijnvg I've added yaml test

martijnvg commented on x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/action/DeleteDataStreamTransportAction.java:
```diff
@@ -106,9 +108,11 @@ static ClusterState removeDataStream(
-        Set<String> dataStreams = new HashSet<>(
-            indexNameExpressionResolver.dataStreamNames(currentState, request.indicesOptions(), request.getNames())
-        );
+        IndicesOptions options = request.indicesOptions();
```
I don't fully understand the purpose of this logic here. It always ensures that expressions get expanded to open data streams / indices. Why is this needed? 

Also the rest spec for this and get data stream api only indicates that `expand_wildcards` is supported, but I think the way the corresponding request class and with this logic also the other indices options rest parameters are supported as well? Is the intention of this logic to only support `expand_wildcards` in get and delete data streams apis?

probakowski replied:
Yes, the intention was to only support `expand_wildcards` but to avoid duplicating parsing parameters code.
`open`/`closed` has no meaning in terms of data stream (we can't close data stream, right?), but if we specify only `hidden` no data stream will be returned (that's the way `IndexNameExpressionResolver` works). This is just convenience for a user.
Other options should be ignored as well by the get/delete data stream APIs so no harm with parsing them.

martijnvg replied:
Ok, I see. Can the logic that rewrites the indices options be extracted in a static helper method with a comment describing what you just explained?

probakowski replied:
Yes, added

martijnvg reviewed (commented):
I left a question about the changes to get and delete data stream APIs.
The rest of the PR looks good!

probakowski committed:
Merge remote-tracking branch 'origin/master' into hidden-data-streams

probakowski committed:
review

probakowski committed:
Merge branch 'master' into hidden-data-streams

probakowski committed:
compilation fix

probakowski committed:
fix javadoc

probakowski committed:
fix javadoc

probakowski commented:
@elasticmachine run elasticsearch-ci/1 (known unrelated failure)

martijnvg reviewed (approved):
LGTM 👍

cjcenizal commented:
Opened https://github.com/elastic/kibana/issues/83709 to track the need to reflect these changes in the UI.
